File: MarkdownEditor/MarkdownEditor/Sources/Converter/ConverterManager.m
//
//  ConverterManager.m
//  MarkdownEditor
//
//  Created by Iwaki Satoshi on 2018/02/27.
//  Copyright © 2018 Satoshi Iwaki and 2026 Vincent Janelle. All rights reserved.
//

#import "ConverterManager.h"
#import "TextConverter.h"
#import "GfmConverter.h"
#import "MarkdownConverter.h"
#import "PhpMarkdownConverter.h"
#import "StrictMarkdownConverter.h"
#import "Logger.h"

NSNotificationName ConverterManagerDidChangeContentNotification = @"ConverterManagerDidChangeContentNotification";

@implementation ConverterManager {
    NSString *_string;
    NSUInteger _selectedConverterIndex;
    NSArray<TextConverter *> *_converters;
}

+ (instancetype)sharedInstance {
    static id sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _converters = @[[[GfmConverter alloc] init],
                        [[MarkdownConverter alloc] init],
                        [[PhpMarkdownConverter alloc] init],
                        [[StrictMarkdownConverter alloc] init],
                        ];
        self.selectedConverterIndex = 0;
    }
    return self;
}

- (NSString *)html {
    @synchronized (self) {
        return self.selectedConverter.html;
    }
}

- (NSArray<NSString *> *)converters {
    @synchronized (self) {
        NSMutableArray *converters = [@[] mutableCopy];
        for (TextConverter *converter in _converters) {
            [converters addObject:converter.title];
        }
        return converters;
    }
}

- (NSUInteger)selectedConverterIndex {
    @synchronized (self) {
        return _selectedConverterIndex;
    }
}

- (void)setSelectedConverterIndex:(NSUInteger)selectedConverterIndex {
    @synchronized (self) {
        _selectedConverterIndex = selectedConverterIndex;
        if (_string) {
            [self reload];
        }
    }
}

- (TextConverter *)selectedConverter {
    @synchronized (self) {
        return _converters[_selectedConverterIndex];
    }
}

- (void)setContentWithString:(NSString *)string {
    @synchronized (self) {
        TextConverter *converter = self.selectedConverter;
        [converter setContentWithString:string];
        LogV(@"*** HTML ***");
        LogV(@"%@", converter.html);
        _string = string;
        [self didChangeContent];
    }
}

- (void)didChangeContent {
    [NSNotificationCenter.defaultCenter postNotificationName:ConverterManagerDidChangeContentNotification
                                                      object:nil];
}

- (void)reload {
    [self setContentWithString:_string];
}

@end

File: MarkdownEditor/MarkdownEditor/Sources/Converter/ConverterManager.swift
import Foundation

extension Notification.Name {
    static let converterManagerDidChangeContent = Notification.Name("ConverterManagerDidChangeContentNotification")
}

@objc(ConverterManager)
class ConverterManager: NSObject {
    @objc(ConverterManagerDidChangeContentNotification)
    static let didChangeContentNotification = Notification.Name.converterManagerDidChangeContent

    @objc(sharedInstance)
    static let shared = ConverterManager()

    private var string: String?
    private let availableConverters: [TextConverter] = [
        GfmConverter(),
        MarkdownConverter(),
        StrictMarkdownConverter()
    ]

    @objc var html: String {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        return selectedConverter.html
    }

    @objc var converters: [String] {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        return availableConverters.map(\.title)
    }

    @objc var selectedConverterIndex: Int = 0 {
        didSet {
            objc_sync_enter(self)
            defer { objc_sync_exit(self) }
            if string != nil {
                reload()
            }
        }
    }

    @objc var selectedConverter: TextConverter {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        return availableConverters[selectedConverterIndex]
    }

    @objc(setContentWithString:)
    func setContent(with string: String) {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        let converter = selectedConverter
        converter.setContent(with: string)
        logVerbose("*** HTML ***")
        logVerbose(converter.html)
        self.string = string
        didChangeContent()
    }

    private func didChangeContent() {
        NotificationCenter.default.post(name: Self.didChangeContentNotification, object: nil)
    }

    private func reload() {
        guard let string else {
            return
        }
        setContent(with: string)
    }
}
